### ISSUE TITLE
Problem with building xmake-repo packages on Windows with clang-cl

### ISSUE THREAD

**Issue by lasagnaphil** [xmake-io/xmake#2211]
### Xmake Version

2.6.4

### Operating System Version and Architecture

Windows 10 21H2

### Describe Bug

Many packages available on xmake-repo can't be compiled on Windows with clang-cl (using `xmake f -p windows --toolchain=clang`), because of the following error:
(This is the error message for the cereal library, but other ones like pugixml and glm also has these errors)

```
'auto' return without trailing return type; deduced return types are a C++14 extension
_NODISCARD constexpr auto _Unfancy(_Ptrty _Ptr) noexcept { // converts from a fancy pointer to a plain pointer
                     ^
In file included from C:\Users\LASAGN~1\AppData\Local\Temp\.xmake\220327\_789D54C1EBAB4A4083A0C58FF085D0B0.cpp:2:
In file included from C:\Users\lasagnaphil\AppData\Local\.xmake\packages\c\cereal\1.3.1\e96e020a73aa4fe4bb5adf34c16f4cff\include\cereal/archives/binary.hpp:32:
In file included from C:\Users\lasagnaphil\AppData\Local\.xmake\packages\c\cereal\1.3.1\e96e020a73aa4fe4bb5adf34c16f4cff\include\cereal/cereal.hpp:33:
In file included from C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.30.30705\include\xstring:14:
In file included from C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.30.30705\include\xmemory:13:
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.30.30705\include\limits:998:5: error: statement not allowed in constexpr function
    do {
    ^
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.30.30705\include\limits:1024:74: error: expected ')'
    _Val = static_cast<_Ty>(_Val - ((_Val >> 1) & static_cast<_Ty>(0x5555'5555'5555'5555ull)));
                                                                         ^
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.30.30705\include\limits:1024:67: note: to match this '('
    _Val = static_cast<_Ty>(_Val - ((_Val >> 1) & static_cast<_Ty>(0x5555'5555'5555'5555ull)));
```

Seems to me that there is a problem with specifying the right C++ version for clang-cl and the MSVC STL library (perhaps some preprocessor variables failed to be set?) Notably, all of the packages which failed to compile had the flag `{languages = "c++11"}` in its build scripts.

### Expected Behavior

The packages compile without any errors.

### Project Configuration

```
add_requires("cereal")

target("helloworld")
    set_kind("binary")
    add_files("main.cpp")
    add_packages("cereal")
```

### Additional Information and Error Logs

I've tried the same config with `set_languages("cxx11")` and `set_languages("cxx17")` and got the same errors.

Note that I've installed Build Tools for Visual Studio along with the Clang compiler, using the installer available from Microsoft (https://visualstudio.microsoft.com/downloads/). I've also registered the path for the clang binaries to PATH.

Also, I've tried executing this in "x64 Native Tools Command Prompt for VS" and had the same errors.

**Comment by waruqi** [comment]
The remote package does not support toolchain switching on the windows platform. You can only use msvc to build packages.

**Comment by lasagnaphil** [comment]
Is this a limitation of xrepo (just unimplemented functionality), or is there a bigger problem with xmake that needs to be resolved? I'm trying to add support for this and it would be great if you could provide some pointers as to what needs to be done.

**Comment by waruqi** [comment]
On the windows platform, all packages and precompiled build packages in the xmake-repo repository are built using msvc by default. github actions also use msvc to test them. It is not reliable if we switch to other compilers.

Right now, I only added toolchain switching support for linux, macos and cross platforms.

https://github.com/xmake-io/xmake/blob/0c51d82b9c4ed8762f11e6a52450c23918cfc49e/xmake/modules/package/tools/cmake.lua#L538-L540

**Comment by lasagnaphil** [comment]
Understood. In the meantime I will use a custom private repo for my clang-cl needs.

By the way, the problem with compiling the cereal package was due to the `on_test` callback failing. Changing the config to `{languages = "c++17"}` instead of `{languages = "c++11"}` fixed the problem for me.

**Comment by waruqi** [comment]
> Understood. In the meantime I will use a custom private repo for my clang-cl needs.
> 
> By the way, the problem with compiling the cereal package was due to the `on_test` callback failing. Changing the config to `{languages = "c++17"}` instead of `{languages = "c++11"}` fixed the problem for me.

you can open a pr to fix package

**Comment by lasagnaphil** [comment]
I'm hesitant about applying this though, since these packages are intended to be compatible with c++11, and the tests should reflect this fact. (And numerous packages other than `cereal` are affected by this change!)

I think the better way to fix this is to explore how to make clang-cl behave well with older language versions (`{languages = "c++11"}`). I suspect that a preprocessor variable isn't being set properly and therefore is using the wrong function definitions (hence the "deduced return types are a C++14 extension" error). I'll keep investigating this problem.

**Comment by lasagnaphil** [comment]
Ah, that might be the problem. From the Clang user manual it says that the flags for setting C++ version is `/std:<version>`instead of `-std=<version>`. Now the complicated part is, the `-Xclang` flag at the beginning does pass the rest of the arguments to the original `clang` engine, so `-Xclang -std=c++11` should work fine in theory. But it seems that this usage doesn't bode well. To show you a quick example, when compiling this simple C++ code:

```c++
#include <limits>

int main(int argc, char* argv[]) {
	return 0;
}
```

When we run `clang-cl -Xclang -std=c++11 main.cpp`, the same error messages happen! I suspect that what `-Xclang -std=c++11` does is forcing c++11 for the clang compiler, but let the standard library use c++17, hence the compile errors. (Note that clang-cl doesn't use libc++ as the std library and uses the MSVC version instead, maybe this is related?) It's weird, but well `clang-cl` is full of dark magic that I don't fully understand yet. 

TLDR: we might need to use `clang-cl /std:<version>` to make this problem go away. I'll fix these flags in `clang_cl.lua` and send a pull request after checking if it works.

**Comment by yangyongzhen** [comment]
I met this error message when use qt quickapp,Does anyone know why？
error: cl : Command line warning D9002 : ignoring unknown option '-std:c++11'
xmake
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2015
checking for Qt SDK directory ... D:/Qt5.12.11/Qt5.12.11/5.12.11/msvc2015_64
checking for Qt SDK version ... 5.12.11
error: cl : Command line warning D9002 : ignoring unknown option '-std:c++11'

**Comment by waruqi** [comment]
> Ah, that might be the problem. From the Clang user manual it says that the flags for setting C++ version is `/std:<version>`instead of `-std=<version>`. Now the complicated part is, the `-Xclang` flag at the beginning does pass the rest of the arguments to the original `clang` engine, so `-Xclang -std=c++11` should work fine in theory. But it seems that this usage doesn't bode well. To show you a quick example, when compiling this simple C++ code:
> 
> ```c++
> #include <limits>
> 
> int main(int argc, char* argv[]) {
> 	return 0;
> }
> ```
> 
> When we run `clang-cl -Xclang -std=c++11 main.cpp`, the same error messages happen! I suspect that what `-Xclang -std=c++11` does is forcing c++11 for the clang compiler, but let the standard library use c++17, hence the compile errors. (Note that clang-cl doesn't use libc++ as the std library and uses the MSVC version instead, maybe this is related?) It's weird, but well `clang-cl` is full of dark magic that I don't fully understand yet.
> 
> TLDR: we might need to use `clang-cl /std:<version>` to make this problem go away. I'll fix these flags in `clang_cl.lua` and send a pull request after checking if it works.

I have improved it, try it again. `xmake update -s dev`

**Comment by lasagnaphil** [comment]
It still gives out the same error after the updates. (By the way, `xmake update -s dev` breaks the installation when you've installed xmake with scoop. so I switched to using the Powershell installation script for now.)

I've also since figured out that using `xmake f --cc=clang-cl --cx=clang-cl` instead of `xmake f --toolchain=clang` removes these compilation errors (To use this in your favorite shell instead of the Visual Studio command prompt, you need to add the LLVM installation to the PATH yourself.) I think it might actually be better to deprecate the custom Windows Clang toolchain in xmake, since clang-cl is intended to be a "switch and forget" replacement for cl.exe. I still don't understand why the error remains under `xmake --toolchain=clang` though. 

But anyways, I've since moved on to not using any remote/local packages and switched to using Git submodules with handmade targets for other technical reasons (more exactly, because I wanted to switch freely between debug/release builds on libraries, and have easy access to the source code. I'll post another issue about some ergonomic issues of the xmake package system soon). So this issue kinda became out of my radar for some time, sorry if I was unresponsive for this Github issue.

**Comment by waruqi** [comment]
> I've also since figured out that using xmake f --cc=clang-cl --cx=clang-cl instead of xmake f --toolchain=clang removes these compilation errors 

so it works now.

> xmake --toolchain=clang 

It is only for clang on linux/macOS and clang on msys2/cygwin in $PATH envirnoment. so it will not use clang-cl.exe

> xmake --toolchain=llvm

It is only for llvm toolchain. xmake will use clang in llvm sdk directory and use llvm-ar/llvm-strip.

**Comment by lasagnaphil** [comment]
> It is only for clang on linux/macOS and clang on msys2/cygwin in $PATH envirnoment. so it will not use clang-cl.exe
It will use clang and ar/strip.

I thought this command would work after seeing it from the blog post (https://tboox.org/2020/06/05/xmake-update-v2.3.4/). Seems that I should have just read the latest docs instead... (https://xmake.io/#/toolchain/builtin_toolchains?id=clang-cl)

Anyways, I think we can close this now, since the issue has now been resolved.

**Comment by waruqi** [comment]
> TLDR: we might need to use clang-cl /std:<version> to make this problem go away. I'll fix these flags in clang_cl.lua and send a pull request after checking if it works.

/std:<version> doesn't always work, but `-Xclang -std=version` does.

https://github.com/xmake-io/xmake/issues/4353#issuecomment-1797159253
